diskquota.c

```
@@ -1365,3 +1518,346 @@ wait_for_worker_new_epoch(PG_FUNCTION_ARGS)
+	{
+		curDB = DiskquotaLauncherShmem->dbArray;
+	}
+	else
+	{
+		curDB++;
+	}
+	for (; curDB < DiskquotaLauncherShmem->dbArrayTail; curDB++)
+	{
+		if (!curDB->in_use) continue;
+		if (curDB->workerId != INVALID_WORKER_ID) continue;
+		if (curDB->dbid == InvalidOid) continue;
+		// TODO: check if it is paused
+		break;
+	}
+	return curDB;
+}
+
+static DiskQuotaWorkerEntry *
+next_worker(void)
+{
+	DiskQuotaWorkerEntry *dq_worker = NULL;
+	dlist_node           *wnode;
+
+	/* acquire worker from worker list */
+	LWLockAcquire(diskquota_locks.workerlist_lock, LW_EXCLUSIVE);
+	if (dlist_is_empty(&DiskquotaLauncherShmem->freeWorkers)) goto out;
+	wnode     = dlist_pop_head_node(&DiskquotaLauncherShmem->freeWorkers);
+	dq_worker = dlist_container(DiskQuotaWorkerEntry, node, wnode);
+	reset_worker(dq_worker);
+	dlist_push_head(&DiskquotaLauncherShmem->runningWorkers, &dq_worker->node);
+	DiskquotaLauncherShmem->running_workers_num++;
+out:
+	LWLockRelease(diskquota_locks.workerlist_lock);
+	return dq_worker;
+}
+
+static void
+wait_bgworker_terminate(BackgroundWorkerHandle *handle)
+{
+	BgwHandleStatus status;
+	int             rc;
+	bool            save_set_latch_on_sigusr1;
+
+	save_set_latch_on_sigusr1 = set_latch_on_sigusr1;
+	set_latch_on_sigusr1      = true;
+	bool timeout              = true;
+
+	PG_TRY();
+	{
+		for (int i = 0; i < 1200; i++)
+		{
+			pid_t pid;
+
+			CHECK_FOR_INTERRUPTS();
+
+			status = GetBackgroundWorkerPid(handle, &pid);
+
+			if (status == BGWH_NOT_YET_STARTED || status == BGWH_STOPPED)
+			{
+				timeout = false;
+				break;
+			}
+
+			rc = WaitLatch(&MyProc->procLatch, WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH, 100L);
+
+			if (rc & WL_POSTMASTER_DEATH)
+			{
+				timeout = false;
+				break;
+			}
+
+			ResetLatch(&MyProc->procLatch);
+		}
+	}
+	PG_CATCH();
+	{
+		set_latch_on_sigusr1 = save_set_latch_on_sigusr1;
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+	if (timeout) elog(WARNING, "Timeout to wait bgworker to terminate");
```

Shall we report an error here? Or warn user to manually killing the bgworker?

---

No, I don't think so. 
Before this commit, diskquota doesn't wait for bgworer to terminate.
I'm wondering if need to wait for the bgworker to terminate, as it increases the deadlock for dropping extension.